nemo_curator/utils/semdedup_utils.py

```diff
@@ -0,0 +1,444 @@
+    return M, M1
+
+
+def get_cluster_reps(
+    cluster_id: int, emb_by_clust_dir: str, id_col: str, sorted_ids: np.ndarray
+) -> torch.Tensor:
+    cluster_i_path = os.path.join(emb_by_clust_dir, f"nearest_cent={cluster_id}")
+    cluster_reps = cudf.read_parquet(
+        cluster_i_path, columns=["embeddings", id_col]
+    ).sort_values(by=id_col)
+    num = cluster_reps.shape[0]
+
+    df_ = pd.DataFrame(
+        {"sorted_ids": sorted_ids, "inverse_sort": list(range(num))}
+    ).sort_values(by="sorted_ids")
+    cluster_reps["inverse_sort_id"] = df_["inverse_sort"].values
+    cluster_reps = cluster_reps.sort_values(by="inverse_sort_id")
+
+    cluster_reps = torch.as_tensor(
+        cluster_reps["embeddings"].list.leaves.values.reshape(len(cluster_reps), -1),
+        device="cuda",
+    )
+    return cluster_reps
+
+
+def get_semantic_matches_per_cluster(
+    cluster_id: int,
+    emb_by_clust_dir: str,
+    sorted_clusters_dir: str,
+    id_col: str,
+    id_col_type: str,
+    eps_list: List[float],
+    output_dir: str,
+    which_to_keep: str,
+) -> None:
+
+    output_df_file_path = os.path.join(output_dir, f"cluster_{cluster_id}.parquet")
+
+    sorted_file = os.path.join(sorted_clusters_dir, f"cluster_{cluster_id}.npy")
+    if not os.path.exists(sorted_file):
+        logging.info(f"{sorted_file} does not exist. Continue")
+        return
+
+    cluster_i = np.load(sorted_file)
+    cluster_size = cluster_i.shape[0]
+    logging.info(f"{cluster_id}: cluster_size: {cluster_size}")
+
+    if cluster_size == 1:
+        points_to_remove_df = pd.DataFrame()
+        points_to_remove_df["indices"] = [0]
+        for eps in eps_list:
+            points_to_remove_df[f"eps={eps}"] = [False]
+        points_to_remove_df.to_parquet(output_df_file_path)
+        return
+
+    clutser_items_indices = list(range(cluster_size))
+
+    which_to_keep = which_to_keep.lower()
+    if which_to_keep == "random":
+        random.shuffle(clutser_items_indices)
+        cluster_i = cluster_i[clutser_items_indices]
+    elif which_to_keep == "easy":
+        clutser_items_indices = clutser_items_indices[::-1]
+        cluster_i = cluster_i[clutser_items_indices]
+
+    text_ids = cluster_i[:, 0].astype(id_col_type)
+
+    cluster_reps = get_cluster_reps(cluster_id, emb_by_clust_dir, id_col, text_ids)
+    M, M1 = _semdedup(cluster_reps, "cuda")
+    assert cluster_reps.shape[0] == len(text_ids)
+
+    M1_id = [text_ids[m] for m in M1]
+
+    points_to_remove_df = cudf.DataFrame()
+    points_to_remove_df["indices"] = clutser_items_indices
+    points_to_remove_df["id"] = text_ids
+    points_to_remove_df["max_id"] = M1_id
+    points_to_remove_df["cosine_sim_score"] = M.numpy().tolist()
+
+    for eps in eps_list:
+        eps_points_to_remove = M > 1 - eps
+        points_to_remove_df[f"eps={eps}"] = eps_points_to_remove
+
+    points_to_remove_df.to_parquet(output_df_file_path)
+    return None
```

Comment:
Nit: delete

Reply:
Resolve by https://github.com/NVIDIA/NeMo-Curator/pull/130/commits/a304629f572c0bb202bd14efa5843b18bb7ca388

Reply:
I think this one is still there

Reply:
Resolved by https://github.com/NVIDIA/NeMo-Curator/pull/130/commits/52480aa4f3dbeb5761a18e5c159591f881dc4a79

Forgot the push earlier.